### DISCUSSION TITLE
Spindle at Speed Questions

### DISCUSSION THREAD

**Discussion by rvalotta** [terjeio/grblHAL#342] [Q&A]
Noticed the other day that the spindle wasn't quite up to speed when starting its cutting. Did some research and found Spindle at speed.  I set it to 6% and started to play with it a little.  I start the spindle and about 5 seconds after I see a alarm (14) and the spindle gets set to off.  I turn off the spindle at speed and it works as expected.

I am using a 15K spindle on a Huanyang VFD being controlled by rs485/modbus.

$0=10.0
$1=255
$2=0
$3=3
$4=0
$5=4
$6=1
$7=1
$10=511
$11=0.020
$12=0.002
$13=0
$14=7
$15=0
$16=0
$17=0
$18=0
$19=0
$20=0
$21=0
$22=35
$23=0
$24=180.0
$25=2500.0
$26=500
$27=5.000
$28=0.100
$29=0.0
$30=24000.000
$31=1000.000
$32=0
$33=5000.0
$34=0.0
$35=6.2
$36=100.0
$37=0
$39=1
$40=1
$43=1
$44=4
$45=3
$46=0
$62=0
$63=2
$64=0
$65=0
$80=1.000
$81=0.010
$82=0.000
$84=0.000
$85=10.000
$90=0.000
$91=0.000
$92=0.000
$95=0.000
$100=400.000
$101=400.000
$102=400.000
$110=2500.000
$111=2500.000
$112=2000.000
$120=150.000
$121=150.000
$122=150.000
$130=1310.000
$131=1260.000
$132=97.000
$170=0.000
$171=0.000
$172=0.000
$340=6.0
$341=1
$342=30.0
$343=25.0
$344=200.0
$345=100.0
$347=5.0
$348=2.500
$349=25.000
$374=3
$375=50
$384=0

**Comment by andrewmarles** [comment]
It looks like the spindle control is using the "spindle_on_delay" variable from the safety door setting when starting the spindle in rpm_controlled mode.  This appears to be hard-coded at the moment to 4 seconds in the SAFETY_DOOR_SPINDLE_DELAY setting.  I think if you uncomment line 92 in config.h you should be able to increase the delay to match your spindle.

There is a $392 setting defined for this variable in the settings structure, but I'm not sure it's used at this moment by the spindle control code.

I have found with the HuanYang that if you enable frequency tracking for the startup (PD025=1) your spindle is able to come up to speed much more quickly.

**Reply by terjeio** [reply]
>  I think if you uncomment line 92 in config.h you should be able to increase the delay to match your spindle.

You will but only if resetting to default settings values with `$RST=*` if the safety door input is not enabled. If enabled `$392` is available, I have to change that to make it available when spindle at speed is supported by the spindle driver...

**Reply by rvalotta** [reply]
I don't have the door safety enabled or compiled in... So I don't think that my issue...

**Reply by terjeio** [reply]
> I don't have the door safety enabled or compiled in... So I don't think that my issue...

When not changed the delay is fixed at 4 seconds from the `SAFETY_DOOR_SPINDLE_DELAY` default. I decided to use that for both the safety door and at speed delay as it is the same thing. To increase the delay you have to do as described above. Alternatively you can set the tolerance back to 0 and use G4 to add a delay for spindle spin up.

`SAFETY_DOOR_SPINDLE_DELAY` should perhaps be renamed to `SPINDLE_SPIN_UP_DELAY`.

Alarm 14 _can_ be due to the VDF not responding, but that is likely not the issue since it works when tolerance is set to 0.

**Reply by rvalotta** [reply]
Can you think of a way to confirm that it is sending and receiving properly to the vfd? More the receiving back.   I mean I know its getting the commands (spindle starts up) but i can't say for a fact that the controller is ever getting anything back.

**Reply by dresco** [reply]
> I know its getting the commands (spindle starts up) but i can't say for a fact that the controller is ever getting anything back.

You can be sure the controller is seeing a response to the spindle control messages, as will go into Alarm state if a reply is not received..

But, as another test, the response to the realtime '?' command will include actual RPM from the VFD. If you send that a few times as it's spinning up, you should see the reported actual speed increasing to match the set speed..

**Reply by rvalotta** [reply]
Thank you for that... Yes the spindle is reporting the speed and yet it is still erroring out when I have spindle at speed turned on.  
Time to dig into the code some more i guess.

**Reply by terjeio** [reply]
> Time to dig into the code some more i guess.

FYI the latest build has `$392` enabled for setting the spin up delay even when safety door input is not enabled.
